torch/csrc/autograd/variable.h

```diff
@@ -1,329 +1,615 @@
-    // variable. These nodes get suppressed in some situations,
-    // see "suppress grad accumulation" below. Note that only variables which
-    // have `requires_grad = True` can have grad accumulators.
+    // interpret the gradient function to be a gradient accumulator, which will
+    // accumulate its inputs into the grad property of the variable. These
+    // nodes get suppressed in some situations, see "suppress gradient
+    // accumulation" below. Note that only variables which have `requires_grad =
+    // True` can have gradient accumulators.
     if (const auto& gradient = grad_fn()) {
       return Edge(gradient, output_nr());
     } else {
       return Edge(grad_accumulator(), 0);
     }
   }
 
-  inline VariableImpl* get() const;
+  /// Returns the input index of the gradient `Function` to which this `Variable`
+  /// is connected.
+  uint32_t output_nr() const noexcept;
 
-  inline const Tensor & data() const;
-  inline       Tensor & data();
+  /// True if this `Variable` is a leaf and thus does not have a `grad_fn`.
+  bool is_leaf() const noexcept;
 
-  inline Tensor opt_data() const;
+  // The Grad Variable
+  //~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 
-  inline const Variable & grad() const;
-  inline       Variable & grad();
+  /// Accesses the gradient `Variable` of this `Variable`.
+  const Variable& grad() const noexcept;
+  Variable& grad() noexcept;
+  void reset_grad() noexcept;
 
-  inline bool is_leaf() const;
+  /// Sets the `requires_grad` property of `Variable`. This should be true for
+  /// leaf variables that want to accumulate gradients, and false for all other
+  /// variables.
+  void set_requires_grad(bool requires_grad) noexcept;
+  bool requires_grad() const noexcept;
 
-  inline const std::shared_ptr<Function>& grad_fn() const;
+  // Versions
+  //~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 
-  // Updates the grad_fn of an existing Variable. Called after in-place modifications.
-  // XXX: this should be called only _after_ the version counter is implemented.
-  inline void rebase_history(int output_nr, std::shared_ptr<Function> grad_fn);
+  /// Increments the version count of this `Variable`.
+  void bump_version() noexcept;
+  void set_version(const VariableVersion& version) noexcept;
 
-  std::shared_ptr<Function> grad_accumulator() const;
+  /// Retrieves this `Variable`s version counter.
+  const VariableVersion& version_counter() const noexcept;
+
+  /// Retrieves the current value of the `Variable`'s version counter.
+  /// Equivalent to calling `version_counter().current_version()`.
+  uint32_t current_version() const noexcept;
+
+  // Autograd Graph Interaction
+  //~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+  /// Update the grad_fn of an existing Variable. Called after in-place
+  /// modifications.
+  void rebase_history(Edge gradient_edge);
```

Comment:
Can you mention this function next to `set_gradient_edge`? It's important that this one is used when in-place modifications are involved